get_info_on_all_users/get_user_role_changes.py

```diff
@@ -0,0 +1,191 @@
+        role_change['before_value'] = user_role_to_tier[role_change['before_value']]
+        if role_change['value'] != role_change['before_value']:
+            tier_changes.append(role_change)
+
+    return tier_changes
+
+def chunk_date_range(args):
+    # Mirror the default of the audit APIs, get records for the last 24 hours
+    since, until = args.since, args.until
+    if not since or not until:
+        now = datetime.now(timezone.utc)
+        yesterday = now - relativedelta.relativedelta(hours=+24)
+        yield (datetime.isoformat(yesterday), datetime.isoformat(now))
+        return
+
+    since = parser.isoparse(since)
+    until = parser.isoparse(until)
+
+    if since > until:
+        raise 'Invalid date range'
+
+    # Audit API requests have a date range limit of 31 days, so we
+    # split the requested date range into 30 day chunks
+    while True:
+        next_since = since + relativedelta.relativedelta(days=+30)
+        if next_since < until:
+            yield (datetime.isoformat(since), datetime.isoformat(next_since))
+            since = next_since
+        else:
+            yield (datetime.isoformat(since), datetime.isoformat(until))
+            return
+
+def main(args, session):
+    user_id, tier_changes = args.user_id, args.tier_changes
+    try:
+        role_changes = []
+        chunked_date_range = chunk_date_range(args)
+        for since, until in chunked_date_range:
+            for record in session.iter_cursor(get_api_path(user_id), params=get_api_params(since, until, user_id)):
+                if args.show_all:
+                    print(json.dumps(record))
+                record_role_changes = get_role_changes(record, args.only_updates)
+                role_changes += record_role_changes
+
+        changes = get_role_tier_changes(role_changes) if tier_changes else role_changes
+        if len(changes):
+            changes = sorted(changes, key=lambda rc: rc['date'])
+            print_changes(changes, tier_changes)
+            if args.filename:
+                write_changes_to_csv(changes, tier_changes, args.filename)
+        else:
+            print(f'No {"tier" if tier_changes else "role"} changes found.')
+
+    except pdpyras.PDClientError as e:
+        print('Could not get user role change audit records')
+        raise e
+
+if __name__ == '__main__':
+    ap = argparse.ArgumentParser(description='Prints all user role or tier changes between the given dates')
+    ap.add_argument('-k', '--api-key', required=True, help='REST API key')
+    ap.add_argument('-s', '--since', required=False, help='Start of date range to search')
+    ap.add_argument('-u', '--until', required=False, help='End of date range to search')
+    ap.add_argument('-i', '--user-id', required=False, help='Filter results to a single user ID')
+    ap.add_argument('-o', '--only-updates', action='store_true', help='Exclude user creates and deletes from role change results')
```

Comment:
I like this option, sounds like this was the original request for this script.

I am glad though were still keeping the create / delete changes. I've seen some fraud activity where users are created and deleted too and I could see why a customer would want to get all the user changes including creates and deletes especially if the user was never actually updated.